plugins/xdebug/xdebug_transforms.cc

```diff
@@ -0,0 +1,158 @@
+/** @file
+
+  Licensed to the Apache Software Foundation (ASF) under one
+  or more contributor license agreements.  See the NOTICE file
+  distributed with this work for additional information
+  regarding copyright ownership.  The ASF licenses this file
+  to you under the Apache License, Version 2.0 (the
+  "License"); you may not use this file except in compliance
+  with the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+#include <limits.h>
+#include <stdio.h>
+#include <string.h>
+#include <functional>
+#include <atomic>
+
+#include "ts/ts.h"
+
+static const std::string_view MultipartBoundary{"\r\n--- ATS xDebug Probe Injection Boundary ---\r\n\r\n"};
+
+struct BodyBuilder {
+  TSVIO output_vio               = nullptr;
+  TSIOBuffer output_buffer       = nullptr;
+  TSIOBufferReader output_reader = nullptr;
+  bool wrote_prebody             = false;
+  bool wrote_body                = false;
+  bool hdr_ready                 = false;
+  std::atomic_flag wrote_postbody;
+
+  int64_t nbytes = 0;
+  TSHttpTxn txn  = nullptr;
+};
+
+static char Hostname[1024];
+
+static std::string
+getPreBody(TSHttpTxn txn)
+{
+  std::stringstream output;
+  output << "{'xDebugProbeAt' : '" << Hostname << "'\n   'captured':[";
+  print_request_headers(txn, output);
+  output << "\n   ]\n}";
+  output << MultipartBoundary;
+  return output.str();
```

Comment:
Isn't this going to make another copy of the bytes?  I would pass in a stringstream to the function.